### PR TITLE
Print activation pspec

### PR DESCRIPTION
**Pull request by NuojCheng**
# Description

This PR improves the observability of sharding constraints in MaxText. When `debug_sharding=True`, the system now logs the JAX type (shape and dtype) and physical partitions for both model weights and activations (with sharding hints). This information is critical for developers designing or debugging complex logical sharding rules. This PR specifically implements the following features:

* **Call-Site Attribution:** Uses `stacklevel` to track and report the exact line number in the model code (e.g., `llama2.py`, `attention_mla.py`) where a sharding constraint is applied, rather than simply reporting the utility function's location.
* **Support for Core Utilities:** Integrated directly into `MaxText.sharding.maybe_shard_with_name` and `MaxText.sharding.maybe_shard_with_logical` (activations won't get printed if using `jax.lax.with_sharding_constraint` or `linen.with_logical_constraint` unfortunately).
* **Trace De-duplication:** Implements a global cache mechanism to prevent log flooding.

In MaxText, JAX tracing occurs three times:
1. **Model Initialization:** Triggered during `train_utils.setup_train_loop` to discover parameter shapes and shardings.
2. **Performance Analysis:** Triggered during `.lower().compile()` to generate the HLO graph for TFLOPs and memory analysis.
3. **First-Step Execution:** The final JIT trace when the first concrete data batch is passed to the compiled train step.

To avoid printing identical sharding information at least three times per tensor, this PR introduces a caching mechanism in `sharding.py`. It stores unique combinations of tensor types and physical partitions, ensuring each unique sharding event is logged exactly once during the first discovery pass.

# Tests

```
alias smoke_train='python3 -m MaxText.train MaxText/configs/base.yml run_name=ncheng-train-base base_output_directory=gs://chengnuojin-maxtext-logs/smoke_train dataset_path=gs://max-datasets-rogue steps=10 log_config=False enable_checkpointing=False enable_goodput_recording=False'
```

Validated on **v5p-8** with `jax==0.8.1`.

## Test 1: DeepSeek-V3 FSDP+EP
**Command:**
`smoke_train debug_sharding=true model_name=deepseek3-test per_device_batch_size=1 ici_expert_parallelism=2`

**Output:** [https://paste.googleplex.com/5196232098185216](https://paste.googleplex.com/5196232098185216)

## Test 2: Llama2-7B FSDP+CP
**Command:**
`smoke_train model_name=llama2-7b per_device_batch_size=1 ici_context_parallelism=2 dataset_type=synthetic debug_sharding=true` 

**Output:** [https://paste.googleplex.com/575225772546464576](https://paste.googleplex.com/5752257725464576)

## Test 3: Default model + Vocab tiling FSDP
**Command:**
`smoke_train debug_sharding=true num_vocab_tiling=4`

**Output:** [https://paste.googleplex.com/4637754698891264](https://paste.googleplex.com/4637754698891264)

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

**Review by gobbleturk — APPROVED** [review]

**Review by richjames0 — APPROVED** [review]
fantastic; lgtm

File: src/MaxText/data_loader.py [hunk 1]
```diff
@@ -20,7 +20,7 @@
 from jax.experimental import checkify
 
 from MaxText import exceptions
```
[the remark this answers is not in the document]
Comment (NuojCheng) [reply src/MaxText/data_loader.py:22]: Not true, ignore
